File: src/main/java/Q543_DiameterOfBinaryTree.java
/***
 * 9mins
 * Runtime: 0 ms, faster than 100.00% of Java online submissions for Diameter of Binary Tree.
 * Memory Usage: 39.5 MB, less than 44.25% of Java online submissions for Diameter of Binary Tree.
 */
public class Q543_DiameterOfBinaryTree {
    int result = 0;
    public int diameterOfBinaryTree(TreeNode root) {
        rec(root);
        return result;
    }

    public int rec(TreeNode root) {
        if(root == null) {
            return 0;
        }
        int left = rec(root.left);
        int right = rec(root.right);
        result = Math.max(result, left + right);
        return Math.max(left, right) + 1;
    }
}

File: LeetCodePractice/LeetCodePractice/Q543_DiameterOfBinaryTree.cpp
#include "LCHeader.h"
#include "TreeNode.cpp"
//7:04 7:09
/*
Runtime: 4 ms, faster than 97.30% of C++ online submissions for Diameter of Binary Tree.
Memory Usage: 20.2 MB, less than 94.78% of C++ online submissions for Diameter of Binary Tree.
*/
class Solution {//7:05
public:
    int max = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        rec(root);
        return max;
    }

    int rec(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        int left = rec(root->left);
        int right = rec(root->right);
        max = std::max(left + right, max);
        return 1 + std::max(left, right);
    }
};